File: src/lib/agent/automation.ts
// Ponte tipada com os comandos das automações.
//
// Automação é um pedido que roda sozinho de tempos em tempos. O agendamento é
// simples de propósito: um intervalo em minutos e a hora da próxima vez em
// epoch. Quem calcula a primeira ocorrência é ESTA camada, que conhece o fuso
// e o relógio da pessoa — o Rust só soma o intervalo.
//
// Fora do Tauri (npm run dev no navegador) tudo é simulado em memória.

import { invoke, isTauri } from "../tauri";
import type { RunMode, RunStatus } from "./types";
import type { WorkMode } from "./scout";

/** Uma automação. Espelha `lr_types::automation::ScheduledTask`. */
export interface ScheduledTask {
  id: string;
  name: string;
  prompt: string;
  workspaceDir: string | null;
  /** Vazio = o primeiro modelo da biblioteca. */
  model: string | null;
  mode: RunMode;
  workMode: WorkMode;
  /** De quanto em quanto tempo, em minutos. `0` = só quando a pessoa mandar. */
  everyMinutes: number;
  /** Próxima execução (epoch ms). */
  nextRunAt: number | null;
  enabled: boolean;
  lastRunAt: number | null;
  lastRunId: string | null;
  lastStatus: RunStatus | null;
  lastSummary: string | null;
  createdAt: number;
}

/** O que a tela manda ao criar ou editar. */
export interface ScheduledTaskInput {
  /** Ausente = criar. */
  id?: string;
  name: string;
  prompt: string;
  workspaceDir?: string | null;
  model?: string | null;
  mode: RunMode;
  workMode: WorkMode;
  everyMinutes: number;
  nextRunAt?: number | null;
  enabled: boolean;
}

/** Intervalos oferecidos na tela, em minutos. */
export const INTERVALS = [0, 60, 60 * 4, 60 * 24, 60 * 24 * 7] as const;

/**
 * Quando esta automação roda pela primeira vez.
 *
 * Para intervalos de um dia ou mais, mira o próximo horário cheio escolhido
 * (padrão: 9h); para os curtos, conta a partir de agora. É aqui que o fuso da
 * pessoa entra na conta — o backend só soma minutos depois disso.
 */
export function firstRun(everyMinutes: number, hour = 9): number | null {
  if (everyMinutes <= 0) return null;
  const now = new Date();
  if (everyMinutes < 60 * 24) return now.getTime() + everyMinutes * 60_000;
  const alvo = new Date(now);
  alvo.setHours(hour, 0, 0, 0);
  if (alvo.getTime() <= now.getTime()) alvo.setDate(alvo.getDate() + 1);
  return alvo.getTime();
}

export function automationsList(): Promise<ScheduledTask[]> {
  if (!isTauri) return Promise.resolve([...mockTasks]);
  return invoke<ScheduledTask[]>("automations_list");
}

export function automationSave(task: ScheduledTaskInput): Promise<ScheduledTask> {
  if (!isTauri) return mockSave(task);
  return invoke<ScheduledTask>("automation_save", { task });
}

export function automationDelete(id: string): Promise<void> {
  if (!isTauri) {
    mockTasks = mockTasks.filter((t) => t.id !== id);
    return Promise.resolve();
  }
  return invoke<void>("automation_delete", { id });
}

/** Roda agora, sem esperar o horário. Devolve o id da execução. */
export function automationRunNow(id: string): Promise<string> {
  if (!isTauri) return mockRunNow(id);
  return invoke<string>("automation_run_now", { id });
}

// ------------------------------------------------------------- simulação ---

let mockTasks: ScheduledTask[] = [
  {
    id: "auto_exemplo",
    name: "Resumo do dia",
    prompt: "Liste o que mudou no projeto hoje e o que ficou pendente.",
    workspaceDir: "/projetos/openweights",
    model: null,
    mode: "approve",
    workMode: "agent",
    everyMinutes: 60 * 24,
    nextRunAt: Date.now() + 3_600_000,
    enabled: true,
    lastRunAt: Date.now() - 86_400_000,
    lastRunId: "run_exemplo",
    lastStatus: "done",
    lastSummary: "3 arquivos alterados, testes passando.",
    createdAt: Date.now() - 7 * 86_400_000,
  },
];

function mockSave(task: ScheduledTaskInput): Promise<ScheduledTask> {
  const anterior = mockTasks.find((t) => t.id === task.id);
  const saved: ScheduledTask = {
    id: task.id ?? `auto_${Date.now()}`,
    name: task.name.trim(),
    prompt: task.prompt.trim(),
    workspaceDir: task.workspaceDir ?? null,
    model: task.model ?? null,
    mode: task.mode,
    workMode: task.workMode,
    everyMinutes: task.everyMinutes,
    nextRunAt: task.nextRunAt ?? null,
    enabled: task.enabled,
    lastRunAt: anterior?.lastRunAt ?? null,
    lastRunId: anterior?.lastRunId ?? null,
    lastStatus: anterior?.lastStatus ?? null,
    lastSummary: anterior?.lastSummary ?? null,
    createdAt: anterior?.createdAt ?? Date.now(),
  };
  mockTasks = anterior
    ? mockTasks.map((t) => (t.id === saved.id ? saved : t))
    : [...mockTasks, saved];
  return Promise.resolve(saved);
}

function mockRunNow(id: string): Promise<string> {
  const runId = `run_${Date.now()}`;
  mockTasks = mockTasks.map((t) =>
    t.id === id
      ? {
          ...t,
          lastRunAt: Date.now(),
          lastRunId: runId,
          lastStatus: "done",
          lastSummary: "Execução simulada.",
        }
      : t,
  );
  return Promise.resolve(runId);
}

File: src-tauri/crates/types/src/automation.rs
//! Automações: um pedido que roda sozinho, de tempos em tempos.
//!
//! A pessoa escreve uma vez ("toda sexta, resuma o que mudou no projeto") e o
//! app cuida do resto — inclusive de subir o motor, porque uma automação que
//! só funciona com o servidor já ligado não automatiza nada.
//!
//! O agendamento é deliberadamente simples: um intervalo em minutos e a hora
//! da próxima vez, em epoch. Quem calcula a primeira ocorrência é a interface,
//! que conhece o fuso e o relógio da pessoa; o Rust só soma o intervalo. Isso
//! evita matemática de fuso horário no backend ao custo de uma hora de desvio
//! quando o horário de verão muda — barato, e visível na tela.

use crate::agent::{RunMode, RunStatus};
use crate::scout::WorkMode;
use serde::{Deserialize, Serialize};

/// Uma tarefa agendada.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledTask {
    pub id: String,
    /// Nome curto, para a pessoa reconhecer na lista.
    pub name: String,
    /// O pedido, igual ao que ela escreveria no chat.
    pub prompt: String,
    pub workspace_dir: Option<String>,
    /// Modelo a usar. Vazio = o que estiver carregado no servidor.
    pub model: Option<String>,
    /// Nível de autorização. Ninguém está olhando quando isto roda: em
    /// `approve` a execução para no primeiro pedido de confirmação e fica
    /// esperando a pessoa — é isso que a tela precisa deixar claro.
    pub mode: RunMode,
    pub work_mode: WorkMode,
    /// De quanto em quanto tempo, em minutos. `0` = só quando ela mandar.
    pub every_minutes: u32,
    /// Quando roda de novo (epoch ms). `None` = não tem próxima.
    pub next_run_at: Option<i64>,
    pub enabled: bool,
    pub last_run_at: Option<i64>,
    pub last_run_id: Option<String>,
    pub last_status: Option<RunStatus>,
    /// Como terminou, em uma linha — ou o motivo de nem ter começado.
    pub last_summary: Option<String>,
    pub created_at: i64,
}

/// O que a interface manda ao criar ou editar (o resto é do backend).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledTaskInput {
    /// Ausente = criar.
    #[serde(default)]
    pub id: Option<String>,
    pub name: String,
    pub prompt: String,
    #[serde(default)]
    pub workspace_dir: Option<String>,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub mode: RunMode,
    #[serde(default)]
    pub work_mode: WorkMode,
    #[serde(default)]
    pub every_minutes: u32,
    /// Primeira vez (epoch ms), calculada pela interface no fuso da pessoa.
    #[serde(default)]
    pub next_run_at: Option<i64>,
    #[serde(default = "yes")]
    pub enabled: bool,
}

fn yes() -> bool {
    true
}

impl ScheduledTaskInput {
    /// Problemas que impedem a tarefa de existir, em linguagem de gente.
    pub fn problem(&self) -> Option<String> {
        if self.name.trim().is_empty() {
            return Some("dê um nome à automação".into());
        }
        if self.prompt.trim().is_empty() {
            return Some("escreva o que a automação deve fazer".into());
        }
        // Um intervalo de poucos minutos com um modelo local vira fila de
        // execuções que nunca alcança o relógio.
        if self.every_minutes > 0 && self.every_minutes < 5 {
            return Some("o intervalo mínimo é de 5 minutos".into());
        }
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn an_automation_needs_a_name_and_something_to_do() {
        let base = ScheduledTaskInput {
            id: None,
            name: "Resumo semanal".into(),
            prompt: "resuma o que mudou".into(),
            workspace_dir: None,
            model: None,
            mode: RunMode::Approve,
            work_mode: WorkMode::Agent,
            every_minutes: 60,
            next_run_at: Some(1),
            enabled: true,
        };
        assert!(base.problem().is_none());

        let sem_nome = ScheduledTaskInput {
            name: "  ".into(),
            ..base.clone()
        };
        assert!(sem_nome.problem().unwrap().contains("nome"));

        let sem_pedido = ScheduledTaskInput {
            prompt: String::new(),
            ..base.clone()
        };
        assert!(sem_pedido.problem().unwrap().contains("fazer"));

        let apressada = ScheduledTaskInput {
            every_minutes: 2,
            ..base.clone()
        };
        assert!(apressada.problem().unwrap().contains("5 minutos"));

        // Sem intervalo é válido: roda só quando a pessoa mandar.
        let manual = ScheduledTaskInput {
            every_minutes: 0,
            ..base
        };
        assert!(manual.problem().is_none());
    }
}
